Create 'Reverse proxy configuration' page(s) on jenkins.io #3204 (#3386)
jenkins-infra/jenkins.io: content/doc/book/system-administration/reverse-proxy-configuration-with-jenkins.adoc

_AllowEncodedSlashes_ is not inherited in Apache configs, so this
directive must be placed inside the _VirtualHost_ definition.

=== Running Jenkins behind Nginx

In situations where you have existing web sites on your server, you may
find it useful to run Jenkins (or the servlet container that Jenkins
runs in) behind http://nginx.org/[Nginx], so that you can bind Jenkins
to the part of a bigger website that you may have. 
This section discusses some of the approaches for doing this.

When a request arrives for certain URLs, Nginx becomes a proxy and
further forward that request to Jenkins, then it forwards the response
back to the client. A typical set up for mod_proxy would look like this:

[source]
----
upstream jenkins {
  keepalive 32; # keepalive connections
  server 127.0.0.1:8080; # jenkins ip and port
}
 
server {
  listen          80;       # Listen on port 80 for IPv4 requests

  server_name     jenkins.example.com;

  #this is the jenkins web root directory (mentioned in the /etc/default/jenkins file)
  root            /var/run/jenkins/war/;

  access_log      /var/log/nginx/jenkins/access.log;
  error_log       /var/log/nginx/jenkins/error.log;

>   ignore_invalid_headers off; #pass through headers from Jenkins which are considered invalid by Nginx server.

Reviewer: Make page formatting work better by reducing line length in examples
```suggestion
 # pass through headers from Jenkins which are considered invalid by Nginx server.
  ignore_invalid_headers off;
```

and the process using them must be owned by root. 
Running Jenkins as root is not recommended - it should be run as its own user. 
One solution is to front Jenkins with a web server such as Apache, and let 
it proxy requests to Jenkins, but this requires maintaining the Apache installation as well.
In situations where you are wanting to run Jenkins on port 80 or 443
(i.e. HTTP/HTTPS), but you do not want to setup a proxy server you can
use _iptables_ on Linux to forward traffic.

==== Ubuntu Installations

The article https://wiki.jenkins.io/display/JENKINS/Installing+Jenkins+on+Ubuntu[Installing
Jenkins on Ubuntu] includes a similar procedure that has been
independently verified to work on Ubuntu 16 LTS, whereas the procedures
shown below did not.

==== Prerequisites

In order to forward traffic from 80/443 to 8080/8443, first you must
ensure that iptables has allowed traffic on all 4 of these ports. 
Use the following command to list the current iptables configuration:

[source]
----
 iptables -L -n
----

You should should see in the output entries for 80, 443, 8080,and 8443.
Here is an example output for comparison.

[source]
----

> ain INPUT (policy ACCEPT)target     prot opt source               destination         Chain INPUT (policy ACCEPT)

Reviewer: Reduce line width by removing empty column from output
```suggestion
ain INPUT (policy ACCEPT)target     prot opt source               destination
```

iptables commands for forwarding ports. 
I found that these did not work for me, and so I researched and 
came up with the information above. 
I do not believe that DNAT is required, as it is typically used to publish a
service from an internal network to a publicly accessible IP. 
But, if the above forwarding commands do not work for you, you might want to
consider trying the commands below. 
Keep in mind that the rules to allow incoming traffic are still required in 
order for these rules to work:

*iptables -t nat -A PREROUTING -i eth0 -p tcp --dport 80 -j DNAT
--to-destination 127.0.0.1:8080*

The first rule uses -j DNAT target of the PREROUTING chain in NAT to
specify a destination IP address and port where incoming packets
requesting a connection to your Jenkins service can be forwarded.

If you have a default policy of DROP in your FORWARD chain, you will
need to add a rule to forward all incoming traffic.

*iptables -A FORWARD -i eth0 -m state --state NEW -m tcp -p tcp -d
127.0.0.1 --dport 8080 -j ACCEPT*

For HTTPS you will need to repeat the above steps but specify port 443
instead of port 80.

==== Using firewalld

> Newer Linux distributions (CentOS 7, RHEL 7, etc.) ship with firewalld
> which serves as a front-end for iptables.  
> Configuration thru firewalld is done via the *firewall-cmd* command.  

Reviewer: Calling CentOS 7 and Red Hat 7 a "newer" Linux distribution is stretching the word "newer" to the breaking point.  Any operating system is not "newer" when it ships command line git 1.8 when many operating systems ship 2.27.
Was newer when the page was written, but not anymore.
```suggestion
Some Linux distributions (CentOS 8, Red hat Enterprise Linux 8, CentOS 7, etc.) ship with firewalld
which serves as a front-end for iptables.
Configuration thru firewalld is done via the *firewall-cmd* command.
```

Instead of using any of the iptables commands mentioned above, 
all you should need to do is something like:

[source]
----

> sudo firewall-cmd --add-port=80/tcp --permanent   (allow incoming connections on port 80.  You can also use --add-service=http instead of adding a port number)
> sudo firewall-cmd --add-port=443/tcp --permanent  (allow incoming connections on port 443.  You can also use --add-service=https instead of adding a port number)
> sudo firewall-cmd --add-forward-port=port=80:proto=tcp:toaddr=127.0.0.1:toport=8080 --permanent

Reviewer: Place comments about the commands and group the operations by port.
```suggestion
# allow incoming connections on port 80.
# You can also use --add-service=http instead of adding a port number
sudo firewall-cmd --add-port=80/tcp --permanent
sudo firewall-cmd --add-forward-port=port=80:proto=tcp:toaddr=127.0.0.1:toport=8080 --permanent

# allow incoming connections on port 443.
# You can also use --add-service=https instead of adding a port number
sudo firewall-cmd --add-port=443/tcp --permanen
t
```

find it useful to run Jenkins (or the servlet container that Jenkins
runs in) behind http://nginx.org/[Nginx], so that you can bind Jenkins
to the part of a bigger website that you may have. 
This section discusses some of the approaches for doing this.

*Make sure that you change the Jenkins httpListenAddress from its
default of 0.0.0.0 to 127.0.0.1 or configure the firewall to block
request on the port Jenkins is bound to, otherwise any IIS-level
restrictions can be easily bypassed by hitting the Jenkins port
directly.*

==== Requirements

* IIS 7.0 or greater.
** IIS 8.5 or greater if you want
https://docs.microsoft.com/en-us/iis/get-started/whats-new-in-iis-85/certificate-rebind-in-iis85[Certificate
Rebind].
* https://www.iis.net/downloads/microsoft/url-rewrite[URL Rewrite 2.1]
or greater.
** As the https://blogs.iis.net/iisteam/url-rewrite-v2-1[announcement]
explains, it introduces a feature flag to turn off the default
non-compliant-RFC3986 behavior. Which is what we want.
* https://www.iis.net/downloads/microsoft/application-request-routing[Application
Request Routing]  3.0 or greater.
* Server access

==== Example use case

> I have a dedicated Jenkins installation on a Windos Server 2012 R2

Reviewer: Spelling fix for a product name
```suggestion
I have a dedicated Jenkins installation on a Windows Server 2012 R2
```

Site_ → __Configuration Editor__ and change the _Section_ to
_system.webServer/rewrite/rules_
. Now you should see the URL Rewrite 2.1 property
_useOriginalURLEncoding_ listed, if not install URL Rewrite 2.1 using
the x86 or x64 installer, not the WebPI one and resume from here after a
reboot.
. Change _useOriginalURLEncoding_ to *False*
.. As the URL Rewrite 2.1 announcement this will change the
value of \{UNENCODED_URL} to make it _RFC3986_ and usable for reverse
proxy forwarding purposes
.. original as in pre 2.1 behaviour. 
. Refresh that tab you were supposed to keep open, or recreate
it.
.. Again, take some time to bask in its glory 
. It should now be white, also the Manage page should no
longer complain!

===== Continue configuring IIS

Some of the things you might want but I won't cover:

* _Hypertext Strict Transport Security_ headers
* Redirecting from non canonical URLs to the canonical URL
(ok, sort of covered this in the web.config example) 
* The X-UA-Compatibility header so that Internet Explorer 11
(or 9, or ...) won't claim to be IE 7 for intranet sites
* Use IIS Crypto to configure cipher suites
* ...

===== A working web.config

*web.config*

[source,xml]

> ----

Reviewer: Could this example have its indentation switched from 4 spaces to 2 spaces and have the longer lines split in a way that is still consistent with valid XML data?

It creates such long examples that the page is unreadable on all but my widest screens

http_access allow localhost
http_access allow manager localhost
http_access allow valid_dst
http_access deny all
http_access deny manager

## mkdir /etc/squid/ssl/ && cd /etc/squid/ssl/ # to generate your self-signed certificate
## openssl genrsa -out jenkins.key 1024
## openssl req -new -key jenkins.key -x509 -out jenkins.crt -days 999
http_port 80 vhost
#https_port 443 cert=/etc/squid/ssl/jenkins.crt key=/etc/squid/ssl/jenkins.key vhost

http_reply_access allow all
icp_access allow all

refresh_pattern -i \.jp(e?g|gif|png|ico)   300  20%  600 override-expire

logformat combined %>a %ui %un \[%tl\] "%rm %ru HTTP/%rv" %Hs %<st "%{Referer}>h" "%{User-Agent}>h" %Ss:%Sh
strip_query_terms off
access_log /var/log/squid/access.log combined

visible_hostname ci.YOUR_DOMAIN
----

This assumes that you run Jenkins on localhost port 8080. 
But you can have it on an other server / different port 
(adjust line starting with cache_peer) +

Of course replace  YOUR_DOMAIN with your domain. +

===== With ssl

Remove one level of comment

[source]
----
 sed s/^#// /etc/squid/squid.conf
----

/!\ If you use the swarm client plugin , the nodes will complain about :

[source]
----

> Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

Reviewer: Split the long line for readability and reduced impact on the rest of the text on the page.
```suggestion
Caused by: sun.security.validator.ValidatorException:
    PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException:
        unable to find valid certification path to requested target
```

See the <<troubleshooting>> section if Jenkins is reporting that your reverse proxy setup is broken or pages are not behaving as expected.

== Configuration Examples

Jenkins works with many different reverse proxies.
This page provides examples for specific reverse proxies, though much of the information also applies to other reverse proxies.

* <<Running Jenkins behind Apache>>
* <<Running Jenkins behind Nginx>>
* <<Running Jenkins behind HAProxy>>
* <<Running Jenkins behind Squid>>
* <<Running Jenkins behind IIS>>

=== Running Jenkins behind Apache

In situations where you have existing web sites on your server,
you may find it useful to run Jenkins (or the servlet container
that Jenkins runs in) behind Apache, so that you can bind Jenkins
to the part of a bigger website that you may have.
This section discusses some of the approaches for doing this.

*Make sure that you change the Jenkins httpListenAddress from its
default of 0.0.0.0 to 127.0.0.1 or any Apache-level restrictions can be
easily bypassed by accessing the Jenkins port directly.*

> 

Reviewer: The Apache configuration sections that appear later in the document seem like they are each offering a different technique that might be used to configure Jenkins behind an Apache reverse proxy.  If that is correct, then I think these should be an introduction paragraph with navigation links that will allow the reader to choose their preferred technique.
```suggestion

There are several different alternatives to configure Jenkins with Apache.
Choose the technique that best meets your needs:

* <<mod_proxy>>
* <<mod_proxy with HTTPS>>
* <<mod_ajp/mod_proxy_ajp>>
* <<mod_proxy_ajp+SSL>>
* <<mod_rewrite>>

```

Double check that all the alternatives are included in the list.  I did not perform the detailed check.

Below is an example of `+ProxyPassMatch+` to proxy all URLs other than
`+/.well-known+` (a URL required by letsencrypt):

[source]
----
ProxyPassMatch  ^/(?\!.well-known)  http://localhost:8080 nocanon
----

The _ProxyRequests Off_ prevents Apache from functioning as a forward
proxy server (except for _ProxyPass_), it is advised to include it
unless the server should function as a proxy.

Both the `+nocanon+` option to `+ProxyPass+`, _and_
`+AllowEncodedSlashes NoDecode+`, are required for certain Jenkins
features to work.

If you are running Apache on a Security-Enhanced Linux (SE-Linux)
machine it is essential to make SE-Linux do the right thing by issuing
as root

[source]
----
setsebool -P httpd_can_network_connect true
----

If this is not issued Apache will not be allowed to forward proxy
requests to Jenkins and only an error message will be displayed.

Because Jenkins already compress its output, you can not use the normal
proxy-html filter to modify urls:

[source]
----
SetOutputFilter proxy-html
----

Instead you can use the following:

[source]
----
SetOutputFilter INFLATE;proxy-html;DEFLATE
ProxyHTMLURLMap http://your_server:8080/jenkins /jenkins
----

> http://wiki.uniformserver.com/index.php/Reverse_Proxy_Server_2:_mod_proxy_html_2 +

Reviewer: I don't understand what is trying to be said here.  This URL opens a page that does not exist on a site that looks somewhat like a wiki.  It either needs to be removed or replaced.

```suggestion
```

But since Jenkins seems to be well behaved it's even better to just not
use SetOutputFilter and ProxyHTMLURLMap.

If there are problems with Jenkins sometimes servicing random garbage
pages, then the following may help:

[source]
----
SetEnv proxy-nokeepalive 1
----

Some plug-ins determine URLs from client requests from Host header, so
if you experience some problems with wrong URLs, you can try to switch
on `+ProxyPreserveHost+` directive, which is switched off by default:

[source]
----
ProxyPreserveHost On
----

===== mod_proxy with HTTPS

> If you'd like to run Jenkins with reverse proxy in HTTPS, one user
> reported that HTTPS needs to be terminated at Jenkins, not at the
> front-end Apache.
> See http://hudson.361315.n4.nabble.com/Hudson-behind-an-Apache-Reverse-Proxy-w-SSL-td370997.html[this
> e-mail thread] for more discussion.
> 
> Note that you also may need to have access to your host via regular
> http, else the admin interface test will report a broken proxy setup.

Reviewer: This is referencing an email conversation from 12 years ago.  It should be removed from the documentation.  There are several blog posts that give good hints that Jenkins and Apache with HTTPS is possible and relatively straighforward.  See:

* https://serverfault.com/questions/653408/correct-apache-reverse-proxy-config-with-ssl-for-jenkins-and-sonar
* https://www.agileana.com/blog/serve-jenkins-over-https-with-apache-as-proxy-and-certbot-letsencrypt-ssl/
* https://www.cloudbooklet.com/how-to-setup-jenkins-with-ssl-with-apache-reverse-proxy-on-ubuntu/
* https://mensfeld.pl/2013/06/jenkins-behind-apache-with-https-proxy-pass-with-ssl/

The oldest of those references is from 7 years ago.  The newest is from only 2 years ago.
```suggestion
```

Alternatively, you can add an additional `+ProxyPassReverse+` directive
to redirect non-SSL URLs generated by Jenkins to the SSL side.
Assuming that your webserver is `+your.host.com+`, placing the following within
the SSL virtual host definition will do the trick:

[source]
----
ProxyRequests     Off
ProxyPreserveHost On
AllowEncodedSlashes NoDecode

<Proxy http://localhost:8081/jenkins*>
  Order deny,allow
  Allow from all
</Proxy>

ProxyPass         /jenkins  http://localhost:8081/jenkins nocanon
ProxyPassReverse  /jenkins  http://localhost:8081/jenkins
ProxyPassReverse  /jenkins  http://your.host.com/jenkins
----

Yet another option is to rewrite the Location headers that contain
non-ssl URL's generated by Jenkins.

> If you want to access Jenkins from https://www.example.com/jenkins,

Reviewer: Escape the URL so that it is not converted into a clickable URL by asciidoc.
```suggestion
If you want to access Jenkins from \https://www.example.com/jenkins,
```

placing the following within the SSL virtual host definition also works:

[source]
----
ProxyRequests     Off
ProxyPreserveHost On
ProxyPass /jenkins/ http://localhost:8081/jenkins/ nocanon
AllowEncodedSlashes NoDecode

<Location /jenkins/>
  ProxyPassReverse /
  Order deny,allow
  Allow from all
</Location>

Header edit Location ^http://www.example.com/jenkins/ https://www.example.com/jenkins/
----

But it may also work fine to just use simple forwarding as above (the
first HTTPS snippet), and add

[source]
----
RequestHeader set X-Forwarded-Proto "https"
RequestHeader set X-Forwarded-Port "443"
----

in the HTTPS site configuration, as the Docker demo (below) does.
(`+X-Forwarded-Port+` is not interpreted by Jenkins prior to
https://issues.jenkins-ci.org/browse/JENKINS-23294[JENKINS-23294] so it
may also be desirable to configure the servlet container to specify the
originating port.)

> The collection of snippets above simply don't work out of the box (July
> 2014), here is a full Apache-oriented "sites-enabled" file (ex:
> "sites-enabled/example") for a dedicated Jenkins host, combining the
> ideas from snippets #1 and #3.
> This was formulated on the TurnKeyLinux Jenkins appliance (v 13.0),
> after having updated Jenkins to "1.572".
> TODO (if anyone understands how to do so): Define a more selective path
> for the <Proxy *> tag, instead of *; I currently have the impression
> that the <Proxy> section is not even needed.

Reviewer: This looks like text that refutes the entire preceding section.  Either this paragraph is correct and the preceding text should be replaced because it does not work or this paragraph should be removed.

Having TODO text in pages we publish is usually best done with either an "in progress" marker (that the page is still evolving) or with a comment for the TODO.  In this case, the TODO appears inside the paragraph.
```suggestion
```

between your Apache and Jenkins (that is, you can't run Jenkins on
http://localhost:8081/ci and have it exposed at http://localhost:80/jenkins)

The _ProxyRequests Off_ prevents Apache from functioning as a forward
proxy server (except for _ProxyPass_), it is advised to include it
unless the server should function as a proxy.

==== Proxying CLI commands with the HTTP(S) transport

Using the plain CLI protocol with the HTTP(S) transport to access
Jenkins through an Apache reverse proxy does not work.
See https://issues.jenkins-ci.org/browse/JENKINS-47279[JENKINS-47279 - Full-duplex HTTP(S) transport with plain CLI protocol does not work with Apache reverse proxy]
for more details.
As a workaround, you can use the link:/doc/book/managing/cli/#ssh[CLI over SSH].

If using Apache check that _nocanon_ is set on _ProxyPass_ and
that _AllowEncodedSlashes_ is set.

_AllowEncodedSlashes_ is not inherited in Apache configs, so this
directive must be placed inside the _VirtualHost_ definition.

=== Running Jenkins behind Nginx

In situations where you have existing web sites on your server, you may
find it useful to run Jenkins (or the servlet container that Jenkins

> runs in) behind http://nginx.org/[Nginx], so that you can bind Jenkins

Reviewer: Let's prefer https rather than http in all URLs.
```suggestion
runs in) behind https://nginx.org/[Nginx], so that you can bind Jenkins
```

      break;
    }
    sendfile on;
  }

  location / {
      sendfile off;
      proxy_pass         http://jenkins;
      proxy_redirect     default;
      proxy_http_version 1.1;

      proxy_set_header   Host              $host;
      proxy_set_header   X-Real-IP         $remote_addr;
      proxy_set_header   X-Forwarded-For   $proxy_add_x_forwarded_for;
      proxy_set_header   X-Forwarded-Proto $scheme;
      proxy_max_temp_file_size 0;

      #this is the maximum upload size
      client_max_body_size       10m;
      client_body_buffer_size    128k;

      proxy_connect_timeout      90;
      proxy_send_timeout         90;
      proxy_read_timeout         90;
      proxy_buffering            off;
      proxy_request_buffering    off; # Required for HTTP CLI commands
      proxy_set_header Connection ""; # Clear for keepalive
  }

}
----

This assumes that you run Jenkins on port 8080.
Remember to create the folder /var/log/nginx/jenkins.

For this set up to work, the context path of Jenkins must be the same
between your Nginx and Jenkins (that is, you can't run Jenkins on
http://localhost:8081/ci and have it exposed at
http://localhost:80/jenkins).

> Set the context path by modifying the hudson.xml configuration file and
> adding --prefix=/hudson (or similar) to the <arguments> entry.

Reviewer: No references to Hudson.  Hudson has been dead for a long time.
```suggestion
Set the context path by modifying the jenkins.xml configuration file and
adding --prefix=/jenkins to the <arguments> entry.
```

If you are having problems with some paths (eg folders) with *Blue
Ocean*, you may need to add the following snippet to your proxy
configuration:

[source]
----
if ($request_uri ~* "/blue(/.*)") {
    proxy_pass http://YOUR_SERVER_IP:YOUR_JENKINS_PORT/blue$1;
    break;
}
----

> To give nginx permission to read Jenkins web root folder, add nginx to

Reviewer: The Nginx project prefers that the name start with a captial 'N' as far as I can tell.
```suggestion
To give Nginx permission to read Jenkins web root folder, add the `nginx` user to
```

If you are experiencing the following error when attempting to run long
CLI commands in Jenkins and Jenkins is running behind Nginx, it
is probably due to Nginx timing out the CLI connection.
You can increase the `+proxy_read_timeout+` setting as necessary so the command will
complete successfully.

[source]
----
WARNING: null
hudson.cli.DiagnosedStreamCorruptionException
Read back: 0x00 0x00 0x00 0x1e 0x07
           'Started reverse-proxy-test #68'
           0x00 0x00 0x00 0x01 0x07 0x0a
Read ahead:
Diagnosis problem:
    java.io.IOException: Premature EOF
        at sun.net.www.http.ChunkedInputStream.readAheadBlocking(ChunkedInputStream.java:565)
        ...
    at hudson.cli.FlightRecorderInputStream.analyzeCrash(FlightRecorderInputStream.java:82)
    at hudson.cli.PlainCLIProtocol$EitherSide$Reader.run(PlainCLIProtocol.java:153)
Caused by: java.io.IOException: Premature EOF
    at sun.net.www.http.ChunkedInputStream.readAheadBlocking(ChunkedInputStream.java:565)
    ...
    at java.io.DataInputStream.readInt(DataInputStream.java:387)
    at hudson.cli.PlainCLIProtocol$EitherSide$Reader.run(PlainCLIProtocol.java:111)
----

> === Forwarding With Iptables

Reviewer: This should be at the same level as "Running behind Nginx" and "Running behind Apache" and should be included in the lists of links at the start of the file.  This is an alternate technique that uses the Linux networking components to transfer traffic from port 80 to port 8080 and from port 443 to 8443.
```suggestion
== Running Jenkins behind iptables
```

I'm not confident how the Linux project would choose to capitalize `iptables` when used in a title.  https://www.linode.com/docs/security/firewalls/control-network-traffic-with-iptables/ uses it with lower case 'i'.  https://help.ubuntu.com/community/IptablesHowTo uses it with capital 'I' in one title and lower case 'i' in another title (in the same article).  Arch Linux uses it as lower case 'i'.

REDIRECT   tcp  --  anywhere             anywhere             tcp dpt:http redir ports 8080
REDIRECT   tcp  --  anywhere             anywhere             tcp dpt:https redir ports 8443

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination
----

 +

Once these rules are set and confirmed with iptables -L -n, and once
your Jenkins instance is up and running on port 8080, attempt to access
your Jenkins instance on port 80 instead of 8080.
It should work and your URL should stay on port 80 - in other words,
it should not get redirected to 8080.
The fact that forwarding from 80 to 8080 (or 443 to 8443) should remain
hidden from the client.

===== Saving iptables Configuration

Using the iptables command to change port configuration and routing
rules only changes the current, in-memory configuration.
It does not persist between restarts of the iptables service.
So, you need to make sure you save the configuration to make the changes permanent.

> Saving the configuration is slightly different between RedHat-based and

Reviewer: Red Hat is always two words
```suggestion
Saving the configuration is slightly different between Red Hat rpm based and
```

Debian-based systems.

> On a RedHat-based system (Fedora, CentOS, RHEL, etc), issue the following command:

Reviewer: ```suggestion
On a Red Hat-based system (Fedora, CentOS, Red Hat Enterprise Linux, Oracle Linux, etc), issue the following command:
```

[source]
----
sudo iptables-save > /etc/sysconfig/iptables
----

On a Debian-based system (Debian, Ubuntu, Mint, etc), issue the
following command:

[source]
----
sudo sh -c "iptables-save > /etc/iptables.rules"
----

The iptables-restore command will need to be executed manually, or your
system configured to automatically run it on boot, against the
/etc/iptables.rules file you have created, in order for your iptables
configuration to be retained across reboots.

> On Ubuntu fastest way is to install iptables-persistent after configuring iptables -
> it will automatically create necessery files from current configuration and load
> them on boot.

Reviewer: Phrasing changes
```suggestion
On Ubuntu, the fastest way is to install `iptables-persistent` after configuring iptables.
It will automatically create the required files from the current configuration and load them on boot.
```

[source]
----
sudo apt-get install iptables-persistent
----

See https://help.ubuntu.com/community/IptablesHowTo for other Ubuntu
options.
There are many other resources describing this; please consult
your system's documentation or search on the internet for information
specific to your flavor of Linux.

If you are unsure at all about what kind of system you have, consult
that system's documentation on how to update iptables configuration.

> ==== Alternate Solution
> 
> Note that an earlier version of this documentation had different
> iptables commands for forwarding ports.
> I found that these did not work for me, and so I researched and
> came up with the information above.
> I do not believe that DNAT is required, as it is typically used to publish a
> service from an internal network to a publicly accessible IP.
> But, if the above forwarding commands do not work for you, you might want to
> consider trying the commands below.
> Keep in mind that the rules to allow incoming traffic are still required in
> order for these rules to work:
> 
> *iptables -t nat -A PREROUTING -i eth0 -p tcp --dport 80 -j DNAT
> --to-destination 127.0.0.1:8080*
> 
> The first rule uses -j DNAT target of the PREROUTING chain in NAT to
> specify a destination IP address and port where incoming packets
> requesting a connection to your Jenkins service can be forwarded.
> 
> If you have a default policy of DROP in your FORWARD chain, you will
> need to add a rule to forward all incoming traffic.
> 
> *iptables -A FORWARD -i eth0 -m state --state NEW -m tcp -p tcp -d
> 127.0.0.1 --dport 8080 -j ACCEPT*
> 
> For HTTPS you will need to repeat the above steps but specify port 443
> instead of port 80.

Reviewer: We should choose one or the other rather than offering an alternate solution.  I don't know which alternate works better, but would rather not offer an alternate solution in our documentation.

If this is the better solution, let's use it and delete the earlier solution.  If this is the poorer solution, then let's remove it from the documentation.
```suggestion
```

. Open the rule you just created
. Under _Conditions_ add:
.. Condition input: *\{CACHE_URL}*
.. Pattern: *^(http|ws)s://*
. Under _Server Variables_ add:
.. Name: *HTTP_FORWARDED*, Value:
*for=\{REMOTE_ADDR};by=\{LOCAL_ADDR};host="\{HTTP_HOST}";proto="https"*,
Replace: yes
... Jenkins runs under Jetty, Jetty supports
https://tools.ietf.org/html/rfc7239[RFC7239], so all should be well.
. Under Action change:
..  Rewrite URL to
*\{C:1}[.nolink]#://jenkins.acme.example:8080\{UNENCODED_URL#}*
... Note that there is no slash between the port number and the opening
curly bracket
.. *Remove* the check from the *Append query string* checkbox
. Apply the changes.
. Edit _C:\Windows\System32\drivers\etc\hosts_ so that
*jenkins.acme.example* points to 127.0.0.1
.. When resolving names Windows will check if the name is its own name
before consulting the hosts file. Meaning that adding _vrtjenkins01_ or
_vrtjenkins01.acme.example_ to the hosts file won't have any effect.
... The hosts file will however be consulted before consulting the DNS
infrastructure

===== Experiencing the dreaded "It appears that your reverse proxy set up is broken." error for yourself

> . [.nolink]#https://jenkins.acme.example/configure#

Reviewer: The confluence style `nolink` markup should be replaced with simpler asciidoc escape.
```suggestion
. \https://jenkins.acme.example/configure
```

. Configure the _Jenkins URL_ to

> be **[.nolink]#https://jenkins.acme.example/#**

Reviewer: Same here and at the other locations with a `[.nolink]` marker
```suggestion
be **[.nolink]#https://jenkins.acme.example/#**
```

Site Request Forgery exploits_
. Go to [.nolink]#https://jenkins.acme.example/manage#
. You will still experience the "It appears that your reverse
proxy set up is broken." as expected
.. If you do not get that at this point, then that is very weird...
Continue anyway.
. Right click the _Configure System_ link and choose to
inspect the element.
.. Make sure you are still on the Manage page as you will want
it as your referrer
. Change the value of the _href_ attribute to be
_administrativeMonitor/hudson.diagnosis.ReverseProxySetupMonitor/test_
. Open the link you just changed in a new tab.
.. Keep this tab open
. Observe the "[.nolink]#https://jenkins.acme.example/manage#
vs http:" error and bask in its glory
.. a white page served with HTTP status code is 200 indicates
all is well
... If you do get that at this point, then that is very
weird... Continue anyway.

===== [.nolink]#Fixing the errors#

. In IIS Manager got to _Application Pools_ then edit
_DefaultAppPool_ so that the _.NET CLR version_ is *No Managed Code*
.. You might find that this is not necessary (at far as you
can tell) for your setup, since IIS will only act as a TLS/SSL

> offloading reverse proxy, we don't need it. IIS on Windows 7 does appear
> to need this to be turned off for some urls
> (https://jenkins.acme.example/*)

Reviewer: No need to describe IIS on Windows 7.  Windows 7 support ended January 14, 2020.
```suggestion
offloading reverse proxy, we don't need it.
```

          </conditions>
          <action type="Redirect"
                  url="{C:1}://jenkins.acme.example{UNENCODED_URL}"
                  appendQueryString="false" redirectType="Permanent" />
        </rule>
        <rule name="ReverseProxyInboundRule1" stopProcessing="true">
          <match url="(.*)" />
          <action type="Rewrite"
                  url="{C:1}://jenkins.acme.example:8080{UNENCODED_URL}"
                  appendQueryString="false" />
          <serverVariables>
            <set name="HTTP_FORWARDED"
                 value="for={REMOTE_ADDR};
                        by={LOCAL_ADDR};
                        host=&quot;{HTTP_HOST}&quot;;
                        proto=&quot;https&quot;" />
          </serverVariables>
          <conditions trackAllCaptures="true">
            <add input="{CACHE_URL}" pattern="^(http|ws)s://" />
            <add input="{HTTP_HOST}" pattern="^jenkins\.acme\.example$" />
          </conditions>
        </rule>
      </rules>
    </rewrite>
    <security>
      <requestFiltering allowDoubleEscaping="true" />
    </security>
  </system.webServer>
</configuration>
----

> 
> ===== Complain about it still not working
> 
> Leave a comment or complain to me over at twitter
> (@Darsstar) +

Reviewer: This does not belong in our documentation
```suggestion
```

the network administrator has set or to access any content from the
Internet which you couldn't also access using your web browser. 
The whole purpose of making Jenkins talk through proxies directly 
through the Internet is convenience. 
For example, Jenkins has a very convenient update mechanism for plugins. 
Any running instance of Jenkins will check the plugin repository on the 
Internet for newly available or updated plugins and allow you to install 
or update a plugin with just a mouse click. 
If Jenkins cannot achieve that because it cannot pass the proxy server and 
cannot reach out on the Internet, 
you could achieve the same by manually downloading the plugins and installing them. 
It means just more work for you._

[[JenkinsBehindProxy-HowJenkinshandlesProxyServers]]
==== How Jenkins handles Proxy Servers

You can configure the proxy server that Jenkins will use by going to
*Manage Jenkins* > *Manage Plugins* > *Advanced*. 
This is preferred over setting JVM properties.

Here's a screenshot showing an example configuration:

image::proxy-configuration.png["Proxy Configuration"]

[cols=",",]
|===
|*Field* |*Description*

> 
> |Server |Host name or address of the proxy server. Example:
> `+proxy.example.com+`
> 
> |Port |The port the proxy server listens on. Example: `+8080+`
> 
> |User name |If the proxy server requires authentication, enter the name
> of the user here.
> 
> |Password |If the proxy server requires authentication, enter the
> password here.
> 
> |No Proxy Host |If certain hosts must be contacted directly (without
> going through the proxy), you can specify their names here. You can even
> use `+*+` as a wildcard. Example: `+*jenkins-test*+``+.example.com+`
> 
> |Test URL |Enter the *full URL* of a host to use for testing the proxy
> configuration. Example: http://jenkins-ci.org/
> |===
> 
> To temporarily disable the proxy configuration, you can add a single
> wildcard character `+*+` on the first line. 
> Erasing the value of *Server* will _discard all values_ and disable 
> the use of a proxy server.
> 
> [[JenkinsBehindProxy-Whichpluginssupportaproxyserver]]
> ===== Which plugins support a proxy server?
> 
> [cols=",,,,",]
> |===
> |*Plugin* |*Since version* |*Supports HTTPS?* |*Supports Exclusions?*
> |*Supports authentication?*
> 
> |https://wiki.jenkins.io/display/JENKINS/Fortify+On+Demand+Uploader+Plugin[Fortify
> On Demand Uploader Plugin]
> |https://github.com/jenkinsci/fortify-on-demand-uploader-plugin/releases/tag/fortify-on-demand-uploader-1.08[1.08]
> |image:../../resources/managing/check.svg[(tick)]
> |image:../../resources/managing/check.svg[(tick)]
> |image:../../resources/managing/check.svg[(tick)]
> 
> |https://wiki.jenkins.io/display/JENKINS/Git+Plugin[Git Plugin]
> |https://github.com/jenkinsci/git-plugin/releases/tag/git-2.1.0[2.1.0]
> |image:../../resources/managing/check.svg[(tick)]
> |image:../../resources/managing/check.svg[(tick)]
> |image:../../resources/managing/check.svg[(tick)]
> 
> |https://wiki.jenkins.io/display/JENKINS/Azure+DevOps+and+Team+Foundation+Server+Plugin[Azure
> DevOps and Team Foundation Server Plugin]
> |https://github.com/jenkinsci/tfs-plugin/releases/tag/tfs-5.1.0[5.1.0]
> |image:../../resources/managing/check.svg[(tick)]
> |image:../../resources/managing/check.svg[(tick)]
> |image:../../resources/managing/check.svg[(tick)]
> 
> |https://wiki.jenkins.io/display/JENKINS/Openid+Connect+Authentication+Plugin[Openid
> Connect Authentication Plugin]
> |https://github.com/jenkinsci/oic-auth-plugin/releases/tag/oic-auth-1.4[1.4]
> |image:../../resources/managing/check.svg[(tick)]
> |image:../../resources/managing/check.svg[(tick)]
> |image:../../resources/managing/check.svg[(tick)]
> |===

Reviewer 2: ```suggestion
```

Reviewer 2: This one is an error from my side in #3204 and it is my fault.

The wiki page https://wiki.jenkins.io/display/JENKINS//JenkinsBehindProxy is in fact another topic.

"Proxy" is for Jenkins to reach web resources outside, while "Reverse Proxy" is for request from the outside to reach Jenkins.

So in fact, this section should be removed from here and deserve a separated page. I'm taking care of the tracking: you only have to remove this (use the suggestion from Github)

Reviewer 2: + do not forget to remove the associated images :)

Author: ok

=== Jenkins says my reverse proxy setup is broken

NOTE: Since Jenkins 1.572 this message can also appear if you don't access
Jenkins through a reverse proxy: Make sure the Jenkins URL configured in
the System Configuration matches the URL you're using to access Jenkins.

> [[Jenkinssaysmyreverseproxysetupisbroken-Symptoms]]

Reviewer 2: ```suggestion
```

==== Symptoms

An error message is displayed in the "Manage Jenkins" page - "It appears

> that your reverse proxy set up is broken"

Reviewer 2: Do you mind moving this block under a subsection "Troubleshooting" , as in the screenshot below?

<img width="313" alt="Screenshot 2020-05-29 at 20 24 59" src="https://user-images.githubusercontent.com/1522731/83292519-868c9200-a1ea-11ea-8bf8-221c8bacb255.png">

> [[Jenkinssaysmyreverseproxysetupisbroken-Background]]

Reviewer 2: ```suggestion
```

==== Background

For a reverse proxy to work correctly, it needs to rewrite both the
request and the response. 
Request rewriting involves receiving an inbound HTTP call and then making 
a forwarding request to Jenkins (sometimes with some HTTP headers modified, sometimes not). 
Failing to configure the request rewriting is easy to catch, because you 
just won't see any pages at all.

But correct reverse proxying also involves one of two options, EITHER

> * rewriting the response (for more information

Reviewer 2: ```suggestion
* rewriting the response
```

> see https://wiki.jenkins.io/display/JENKINS/Hyperlinks+in+HTML[Hyperlinks in HTML]). 

Reviewer 2: ```suggestion
```

Reviewer 2: because this link points to the wiki and it does not add a lot of useful info for the reader

The primary place where this needs to happen is the
"Location" header in the response, which is used during redirects.

> Jenkins will send back "Location:

Reviewer 2: ```suggestion
Jenkins will send back the header `Location: http://actual.server:8080/jenkins/foobar`
```

> http://actual.server:8080/jenkins/foobar" and the reverse proxy needs to

Reviewer 2: ```suggestion
and the reverse proxy needs to
```

> rewrite this to "Location: http://nice.name/jenkins/foobar".

Reviewer 2: ```suggestion
rewrite this to `Location: http://nice.name/jenkins/foobar`.
```

warnings. 

Be sure to set the `+X-Forwarded-Proto+` header if your reverse proxy is
accessed via HTTPS and then Jenkins itself is accessed via HTTP i.e.
proxying HTTPS to HTTP.

Changing the context path of Jenkins with a reverse proxy is fraught with danger. 
There are lots of URLs that you need to rewrite correctly,
and even if you get the ones in HTML files you may miss some in
javascript, CSS or XML resources.

The recommendation is to ensure that Jenkins is running at the context
path that your reverse proxy is serving Jenkins at. 
You will have the least pain if you keep to this principle.

While it is technically possible to use rewrite rules to change the context path, 
you should be aware that it would be a lot of work to find and fix everything in 
your rewrite rules and the reverse proxy will spend most of its time rewriting
responses from Jenkins. 
Much easier to change Jenkins to run at the context path your reverse proxy is
expecting, e.g. if your reverse proxy is forwarding requests at
https://manchu.example.org/foobar/ to Jenkins then you could just use
`+java -jar jenkins.war --prefix /foobar+` to start jenkins using
`+/foobar+` as the context path

> [[Jenkinssaysmyreverseproxysetupisbroken-FurtherDiagnosis]]

Reviewer 2: ```suggestion
```

==== Further Diagnosis

For further diagnosis, try using cURL:

[source,sh]
----
curl -iL -e http://your.reverse.proxy/jenkins/manage \
   http://your.reverse.proxy/jenkins/administrativeMonitor/hudson.diagnosis.ReverseProxySetupMonitor/test
----

(assuming your Jenkins is located at

> http://your.reverse.proxy/jenkins/ - and is open to anonymous read

Reviewer 2: ```suggestion
`http://your.reverse.proxy/jenkins/` - and is open to anonymous read
```

access)

> [[Jenkinssaysmyreverseproxysetupisbroken-ServerConfigurationGuides]]

Reviewer 2: ```suggestion
```

==== Server Configuration Guides

While the pages talk primarily about Apache / NGinX / HAProxy / Squid,
they also have information that applies to other reverse proxies.

* <<Running Jenkins behind Apache>>
* <<Running Jenkins behind Nginx>>
* <<Running Jenkins behind HAProxy>>
* <<Running Jenkins behind Squid>>
* <<Running Jenkins behind IIS>>

If using Apache check that _nocanon_ is set on _ProxyPass_ and
that _AllowEncodedSlashes_ is set as per the Apache link above.

_AllowEncodedSlashes_ is not inherited in Apache configs, so this

> directive must be placed inside the _VirtualHost_ definition.

Reviewer 2: Could you move this elements in the "Running Jenkins behind Apache" section please?

== Configuration Examples

=== Running Jenkins behind Apache

In situations where you have existing web sites on your server, 
you may find it useful to run Jenkins (or the servlet container
that Jenkins runs in) behind Apache, so that you can bind Jenkins 
to the part of a bigger website that you may have. 
This section discusses some of the approaches for doing this.

*Make sure that you change the Jenkins httpListenAddress from its
default of 0.0.0.0 to 127.0.0.1 or any Apache-level restrictions can be
easily bypassed by hitting the Jenkins port directly.*

> [[RunningJenkinsbehindApache-mod_proxy]]

Reviewer 2: ```suggestion
```

==== mod_proxy

http://httpd.apache.org/docs/2.0/mod/mod_proxy.html[mod_proxy] works by
making Apache perform "reverse proxy" — when a request arrives for
certain URLs, Apache becomes a proxy and further forward that request to
Jenkins, then it forwards the response back to the client.

The following Apache modules must be installed :

> ....

Reviewer 2: ```suggestion
[source]
----
```

a2enmod proxy
a2enmod proxy_http
a2enmod headers

> ....

Reviewer 2: ```suggestion
----
```

A typical set up for mod_proxy would look like this:

> ....

Reviewer 2: ```suggestion
[source]
----
```

ProxyPass         /jenkins  http://localhost:8081/jenkins nocanon
ProxyPassReverse  /jenkins  http://localhost:8081/jenkins
ProxyRequests     Off
AllowEncodedSlashes NoDecode

# Local reverse proxy authorization override
# Most unix distribution deny proxy by default (ie /etc/apache2/mods-enabled/proxy.conf in Ubuntu)
<Proxy http://localhost:8081/jenkins*>
  Order deny,allow
  Allow from all
</Proxy>

> ....

Reviewer 2: ```suggestion
----
```

This assumes that you run Jenkins on port 8081.

For this set up to work, the context path of Jenkins *must be the same*
between your Apache and Jenkins (that is, you can't run Jenkins on
http://localhost:8081/ci and have it exposed at
http://localhost:80/jenkins). +
Set the context path in Windows by modifying the `+jenkins.xml+`
configuration file and adding --prefix=/jenkins (or similar) to the
<arguments> entry. +
Set the context path  when using the Ubuntu package by adding
`+--prefix=/jenkins+` to `+JENKINS_ARGS+` in `+/etc/default/jenkins+` (
or in `+/etc/sysconfig/jenkins+` for RHEL/CentOS package)

When running on a dedicated server and you are using / as context, make
sure you add a slash at the end of all URLs in proxy params in apache.
Otherwise you might run into proxy errors. So

> ....

Reviewer 2: ```suggestion
[source]
----
```

ProxyPass / http://localhost:8080/ nocanon

> ....

Reviewer 2: ```suggestion
----
```
